File: ninloader/python/ninloader/core/refresh.py
"""Token refresh dispatcher — routes refresh requests to the appropriate collector."""

from __future__ import annotations

from typing import Optional

from ..types import CollectedToken, RefreshResult
from ..collectors import CollectorRegistry


def refresh_token(token: CollectedToken) -> RefreshResult:
    """Attempt to refresh a collected token using the appropriate collector.

    Looks up the collector by (service, source) and delegates to its refresh().
    """
    collector_cls = CollectorRegistry.get(token.service, token.source)
    if not collector_cls:
        return RefreshResult(
            success=False,
            service=token.service,
            source=token.source,
            error=f"No collector found for {token.service}/{token.source}",
        )

    collector = collector_cls()

    if not collector.is_platform_supported():
        return RefreshResult(
            success=False,
            service=token.service,
            source=token.source,
            error=f"Collector {token.service}/{token.source} not supported on this platform",
        )

    try:
        return collector.refresh(token)
    except Exception as e:
        return RefreshResult(
            success=False,
            service=token.service,
            source=token.source,
            error=str(e),
        )

File: ninloader/go/internal/refresh/refresh.go
// Package refresh dispatches token refresh requests to the appropriate collector.
package refresh

import (
	"github.com/ninken/ninloader-go/internal/registry"
	"github.com/ninken/ninloader-go/internal/types"
)

// RefreshToken attempts to refresh a collected token using the appropriate collector.
// It looks up the collector by (service, source) and delegates to its Refresh() method.
func RefreshToken(token *types.CollectedToken) *types.RefreshResult {
	c := registry.Get(token.Service, token.Source)
	if c == nil {
		return &types.RefreshResult{
			Success: false,
			Service: token.Service,
			Source:  token.Source,
			Error:   "no collector found for " + token.Service + "/" + token.Source,
		}
	}

	if !c.IsPlatformSupported() {
		return &types.RefreshResult{
			Success: false,
			Service: token.Service,
			Source:  token.Source,
			Error:   "collector " + token.Service + "/" + token.Source + " not supported on this platform",
		}
	}

	defer func() {
		// If the collector's Refresh() panics, we return an error result.
		// The named return is handled by the outer function; this is a safety net.
	}()

	return safeRefresh(c, token)
}

// safeRefresh calls Refresh() on a collector, recovering from panics.
func safeRefresh(c interface {
	Refresh(token *types.CollectedToken) *types.RefreshResult
}, token *types.CollectedToken) (result *types.RefreshResult) {
	defer func() {
		if r := recover(); r != nil {
			result = &types.RefreshResult{
				Success: false,
				Service: token.Service,
				Source:  token.Source,
				Error:   "panic during refresh: " + formatPanic(r),
			}
		}
	}()
	return c.Refresh(token)
}

// formatPanic converts a recovered panic value to a string.
func formatPanic(r any) string {
	switch v := r.(type) {
	case string:
		return v
	case error:
		return v.Error()
	default:
		return "unknown panic"
	}
}
